blog/content/post/2020-01-23-vmx-exit-reasons-and-handlers.md

```diff
@@ -0,0 +1,475 @@
+Prints general-purpose registers, control registers, guest address (both linear
+and physical), exit reason and exit qualification.
+
+```
+vcpu::add_exit_handler(const handler_delegate_t &d)
+```
+
+> Adds an exit function to the exit list. Exit functions are executed
+> right after a vCPU exits for any reason. Use this with care because
+> this function will be executed a lot.
+>
+> Note the return value of the delegate is ignored
+
+More about delegates below.
+
+```
+vcpu::add_handler(::intel_x64::vmcs::value_type reason,const handler_delegate_t &d)
+```
+
+> Adds an exit handler to the vCPU
+
+Generic way of adding handlers for all defined reasons. Bareflank's names for
+exit reasons can be found in [32bit_read_only_data_fields.h](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfintrinsics/include/arch/intel_x64/vmcs/32bit_read_only_data_fields.h#L155).
+For most common exit reasons (or those requiring some additional work) there
+are specialized `add_*_handler()` methods.
+
+
+```
+vcpu::add_io_instruction_handler(
+    vmcs_n::value_type port,
+    const io_instruction_handler::handler_delegate_t &in_d,
+    const io_instruction_handler::handler_delegate_t &out_d)
+```
+
+This method, apart from adding a handler for exit reason, also enables exiting
+on given port by setting the appropriate bit in I/O bitmaps. Separate handlers
+can be defined for read and write operations. Notice that it uses a different
+type for handler delegate - more about it later.
+
+```
+vcpu::add_default_io_instruction_handler(const ::handler_delegate_t &d)
+```
+
+Nothing special about this particular adder, I just listed it to show that some
+exit reasons have default handlers. Those are called if all other handlers
+returned `false`.
+
+```
+vcpu::trap_on_msr_access(vmcs_n::value_type msr)
+```
+
+> Sets a '1' in the MSR bitmap corresponding with the provided msr. All
+> attempts made by the guest to read/write from the provided msr will be
+> trapped by the hypervisor.
+
+Helper function that enables exiting without installing a new handler. Might be
+used along with `pass_through_msr_access()` to toggle exiting. If run without
+installing a handler, the default handler is used.
+
+### Handlers, delegates and info
+
+VM exit lands in [exit_handler_entry.asm](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/src/hve/arch/intel_x64/exit_handler_entry.asm)
+initially, where it just saves guest values of registers and calls
+`exit_handler::handle()` from [exit_handler.cpp](https://github.com/Bareflank/hypervisor/blob/ba613e2c687f7042bac6886858cf6da3132a61d6/bfvmm/src/hve/arch/intel_x64/exit_handler.cpp#L727).
+At the point of merging EAPIs to the main tree, this was done with some
+assumptions about name mangling, it is no longer the case in newer code.
+
+Apart from exception handling and dumping CPU state in case of not handled exit
+reason, this function performs two loops:
+
+```
```

Comment:
```suggestion
```c
```

Reply:
@artur-rs I'm not sure what is supported by Hugo, but we should have information in README how given code snippets should be marked, good looking code is easier to read.

Reply:
@artur-rs maybe there is marker for C++?

Reply:
@pietrushnic https://gohugo.io/content-management/syntax-highlighting/#list-of-chroma-highlighting-languages

Markers for C++: 
```
C, CPP, H, c++, cc, cp, cpp, cxx, h++, hh, hpp, hxx
```

I will add this to README as well

Reply:
All markers added